File: SOA/app/src/main/java/com/example/soa/MainActivity.java
package com.example.soa;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.util.Log;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.FirebaseApp;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {

    private ArrayList<Items> itemsArrayList;
    private RecyclerView recyclerView;
    private CustomAdapter adapter;
    private FirebaseFirestore db = FirebaseFirestore.getInstance();
    private String TAG = "PESAN";
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        getData();

        recyclerView = (RecyclerView) findViewById(R.id.recycler_view);

        adapter = new CustomAdapter(itemsArrayList);

        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(MainActivity.this);

        recyclerView.setLayoutManager(layoutManager);

        recyclerView.setAdapter(adapter);

    }

    void getData(){
        itemsArrayList = new ArrayList<>();
        db.collection("inventory")
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()) {
                            for (QueryDocumentSnapshot document : task.getResult()) {
                                itemsArrayList.add(new Items(document.getData().get("name").toString(),
                                        Integer.parseInt(document.getData().get("price").toString()),
                                        Integer.parseInt(document.getData().get("quantity").toString()),
                                        document.getData().get("unit").toString()));
                                adapter.notifyDataSetChanged();
                            }
                        } else {
                            Log.w(TAG, "Error getting documents.", task.getException());
                        }
                    }
                });

    }
}

File: SOAKotlin/app/src/main/java/com/example/soa_kotlin/MainActivity.kt
package com.example.soa_kotlin

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase

class MainActivity : AppCompatActivity() {
    private lateinit var itemList:ArrayList<Items>
    private lateinit var adapter: CustomAdapter
    val TAG = "PESAN"
    val db = Firebase.firestore
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        getData()
        val recyclerView: RecyclerView = findViewById(R.id.recycler_view)
        adapter = CustomAdapter(itemList)
        val layoutManager = LinearLayoutManager(applicationContext)
        recyclerView.layoutManager = layoutManager
        recyclerView.adapter = adapter
    }
    private fun getData(){
        itemList = ArrayList()
        db.collection("inventory")
            .get()
            .addOnSuccessListener { result ->
                for (document in result) {
                    itemList.add(
                        Items(
                            document.data["name"].toString(),
                            document.data["price"].toString().toInt(),
                            document.data["quantity"].toString().toInt(),
                            document.data["unit"].toString()
                        )
                    )
                    adapter.notifyDataSetChanged()
                }
            }
            .addOnFailureListener { exception ->
                Log.d(TAG, "Error getting documents: ", exception)
            }
    }
}